sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -82,31 +83,61 @@
  */
 public final class BatchSpanProcessor implements SpanProcessor {
 
+  private static final Logger logger = Logger.getLogger(BatchSpanProcessor.class.getName());
+
   private static final String WORKER_THREAD_NAME =
       BatchSpanProcessor.class.getSimpleName() + "_WorkerThread";
-  private static final String TIMER_THREAD_NAME =
-      BatchSpanProcessor.class.getSimpleName() + "_TimerThread";
-  private final Worker worker;
-  private final Thread workerThread;
+
+  private static final BoundLongCounter droppedSpans;
+
+  static {
+    Meter meter = OpenTelemetry.getMeter("io.opentelemetry.sdk.trace");
+    LongCounter droppedSpansCounter =
+        meter
+            .longCounterBuilder("droppedSpans")
+            .setUnit("1")
+            .setDescription(
+                "The number of spans dropped by the BatchSpanProcessor due to high throughput.")
+            .build();
+    droppedSpans =
+        droppedSpansCounter.bind(
+            Labels.of("spanProcessorType", BatchSpanProcessor.class.getSimpleName()));
+  }
+
+  private final int maxExportBatchSize;
+  private final long scheduleDelayMillis;
+  private final int exporterTimeoutMillis;
   private final boolean sampled;
 
+  private final SpanExporter spanExporter;
+  private final BlockingQueue<ReadableSpan> queue;
+  private final ScheduledExecutorService executor;
+
+  private final List<ReadableSpan> spanBuffer;
+
+  private volatile boolean closed;
+  private volatile ScheduledFuture<?> nextExport;
+
   private BatchSpanProcessor(
       SpanExporter spanExporter,
       boolean sampled,
       long scheduleDelayMillis,
       int maxQueueSize,
       int maxExportBatchSize,
       int exporterTimeoutMillis) {
-    this.worker =
-        new Worker(
-            spanExporter,
-            scheduleDelayMillis,
-            maxQueueSize,
-            maxExportBatchSize,
-            exporterTimeoutMillis);
-    this.workerThread = new DaemonThreadFactory(WORKER_THREAD_NAME).newThread(worker);
-    this.workerThread.start();
+    this.maxExportBatchSize = maxExportBatchSize;
+    this.scheduleDelayMillis = scheduleDelayMillis;
+    this.exporterTimeoutMillis = exporterTimeoutMillis;
     this.sampled = sampled;
+
+    this.spanExporter = spanExporter;
+    queue = new ArrayBlockingQueue<>(maxQueueSize);
+    executor =
+        Executors.newSingleThreadScheduledExecutor(new DaemonThreadFactory(WORKER_THREAD_NAME));
```

Comment:
Can we pass in the executor here so that the BSP can share resources and leave the executor as an outside concern?

Reply:
I don't think there should be a fallback executor. The executor should always be an external concern IMHO. Users should be encouraged to think about these things... :-)